app/scripts/controllers/network/network-controller.ts

```diff
@@ -0,0 +1,1171 @@
+ * for a non-Infura network, then whether the network is being operated locally
+ * or remotely. Primarily used to build the network client and check the
+ * availability of a network.
+ */
+type ProviderType = BuiltInInfuraNetwork | typeof NETWORK_TYPES.RPC;
+
+/**
+ * The network ID of a network.
+ */
+type NetworkId = `${number}`;
+
+/**
+ * The ID of a network configuration.
+ */
+type NetworkConfigurationId = string;
+
+/**
+ * The chain ID of a network.
+ */
+type ChainId = Hex;
+
+/**
+ * The set of event types that NetworkController can publish via its messenger.
+ */
+export enum NetworkControllerEventType {
+  /**
+   * @see {@link NetworkControllerNetworkWillChangeEvent}
+   */
+  NetworkWillChange = 'NetworkController:networkWillChange',
+  /**
+   * @see {@link NetworkControllerNetworkDidChangeEvent}
+   */
+  NetworkDidChange = 'NetworkController:networkDidChange',
+  /**
+   * @see {@link NetworkControllerInfuraIsBlockedEvent}
+   */
+  InfuraIsBlocked = 'NetworkController:infuraIsBlocked',
+  /**
+   * @see {@link NetworkControllerInfuraIsUnblockedEvent}
+   */
+  InfuraIsUnblocked = 'NetworkController:infuraIsUnblocked',
+}
+
+/**
+ * `networkWillChange` is published when the current network is about to be
+ * switched, but the new provider has not been created and no state changes have
+ * occurred yet.
+ */
+type NetworkControllerNetworkWillChangeEvent = {
+  type: NetworkControllerEventType.NetworkWillChange;
+  payload: [];
+};
+
+/**
+ * `networkDidChange` is published after a provider has been created for a newly
+ * switched network (but before the network has been confirmed to be available).
+ */
+type NetworkControllerNetworkDidChangeEvent = {
+  type: NetworkControllerEventType.NetworkDidChange;
+  payload: [];
+};
+
+/**
+ * `infuraIsBlocked` is published after the network is switched to an Infura
+ * network, but when Infura returns an error blocking the user based on their
+ * location.
+ */
+type NetworkControllerInfuraIsBlockedEvent = {
+  type: NetworkControllerEventType.InfuraIsBlocked;
+  payload: [];
+};
+
+/**
+ * `infuraIsBlocked` is published either after the network is switched to an
+ * Infura network and Infura does not return an error blocking the user based on
+ * their location, or the network is switched to a non-Infura network.
+ */
+type NetworkControllerInfuraIsUnblockedEvent = {
+  type: NetworkControllerEventType.InfuraIsUnblocked;
+  payload: [];
+};
+
+/**
+ * The set of events that the NetworkController messenger can publish.
+ */
+type NetworkControllerEvent =
+  | NetworkControllerNetworkDidChangeEvent
+  | NetworkControllerNetworkWillChangeEvent
+  | NetworkControllerInfuraIsBlockedEvent
+  | NetworkControllerInfuraIsUnblockedEvent;
+
+/**
+ * The messenger that the NetworkController uses to publish events.
+ */
+type NetworkControllerMessenger = RestrictedControllerMessenger<
+  typeof name,
+  never,
+  NetworkControllerEvent,
+  never,
+  NetworkControllerEventType
```

Comment:
We don't need to list the internal events here, they're allowed by default. This generic parameter is just meant for _external_ events that are allowed.